community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsTracker.java

```diff
@@ -17,282 +17,431 @@
-    private volatile CountsTrackerState state;
 
-    public CountsTracker( StringLogger logger, FileSystemAbstraction fs, PageCache pageCache, File storeFileBase )
+    public CountsTracker( StringLogger logger, FileSystemAbstraction fs, PageCache pages, File baseFile )
     {
+        super( fs, pages, baseFile, 16, 16, Meta.KEYS );
         this.logger = logger;
-        this.alphaFile = storeFile( storeFileBase, ALPHA );
-        this.betaFile = storeFile( storeFileBase, BETA );
-        CountsStore store = openStore( logger, fs, pageCache, this.alphaFile, this.betaFile );
-        this.state = new ConcurrentCountsTrackerState( store );
     }
 
-    private static CountsStore openStore( StringLogger logger, FileSystemAbstraction fs, PageCache pageCache,
-                                          File alphaFile, File betaFile )
+    @Override
+    protected Meta initialMetadata()
     {
-        try
-        {
-            if ( !fs.fileExists( alphaFile ) || !fs.fileExists( betaFile ) )
-            {
-                throw new UnderlyingStorageException(
-                        "Expected both counts store files " + alphaFile + " and " + betaFile + " to exist. You may " +
-                        "recreate the counts store by shutting down the database first, deleting the counts store " +
-                        "files manually and then restarting the database" );
-            }
+        return new Meta( -1, 1 );
+    }
 
-            CountsStore alphaStore = openVerifiedCountsStore( fs, pageCache, alphaFile );
-            CountsStore betaStore = openVerifiedCountsStore( fs, pageCache, betaFile );
+    @Override
+    protected int compareMetadata( Meta lhs, Meta rhs )
+    {
+        return compare( lhs, rhs );
+    }
 
-            boolean isAlphaCorrupted = alphaStore == null;
-            boolean isBetaCorrupted = betaStore == null;
-            if ( isAlphaCorrupted && isBetaCorrupted )
-            {
-                throw new UnderlyingStorageException(
-                        "Neither of the two store files could be properly opened. Please shut down the database and " +
-                        "delete " + alphaFile + " and " + betaFile + " to have the database recreate the counts" +
-                        "store on next startup" );
-            }
+    static int compare( Meta lhs, Meta rhs )
+    {
+        int cmp = Long.compare( lhs.txId, rhs.txId );
+        if ( cmp == 0 )
+        {
+            cmp = Long.compare( lhs.minorVersion, rhs.minorVersion );
+        }
+        return cmp;
+    }
 
-            if ( isAlphaCorrupted )
-            {
-                logger.debug( "CountsStore picked " + betaFile + " since " + alphaFile + " could not be opened " +
-                              "(txId=" + betaStore.lastTxId() + ", minorVersion=" + betaStore.minorVersion() + ")" );
-                return betaStore;
-            }
+    @Override
+    protected Meta updateMetadata( Meta metadata, Meta.Diff changes )
```

Comment:
Because the `AbstractKeyValueStore` does not know about `Meta`, the point is that the "framework" should not know about the format of the metadata of the specific implementation.